### ISSUE TITLE
early stopping

### ISSUE THREAD

**Issue by zacwellmer** [endernewton/tf-faster-rcnn#28]
has anyone considered building early stopping into the training process?

**Comment by endernewton** [comment]
what do you mean by early stopping? you can just set the ITERSIZE to some smaller value.

**Comment by zacwellmer** [comment]
right but the point of early stopping isn't necessarily to stop at a small itersize value but rather to avoid from overfitting to the training set. So for example, if during the training process at every snapshot we tested on either the test/validation set we would be able to know when our model is being overfit. In the config we could set something that halts the training process after k snapshots of increasing test error.

**Comment by endernewton** [comment]
I see. This is a very nice pointer! I am currently in the debugging stage
for resnet now so I may not be able to help with that for the moment. You
are welcome to try it and push a pull request!

The code does evaluate the validation set from time to time to plot a loss
curve, but it is noisy because it only evaluates 1 image at a time. Also so
far, I do find the network either saturates at a performance and does not
improve further (can have some fluctuation), or benefits from longer
training (COCO), hasn't really tried early stopping yet.

---------------------
Ender X. Chen

Language Technologies Institute,
School of Computer Science,
Carnegie Mellon University, Pittsburgh, PA

**Comment by zacwellmer** [comment]
I'll work on implementing this. As far as the best way to do it: do you think it would be best to handle this from trainval_net.py? because attempting to implement in lib/model/train_val.py will make things a bit hairy since our network changes between training and testing

**Comment by endernewton** [comment]
Maybe you can write a separate train_val_es.py in model/?

---------------------
Ender X. Chen

Language Technologies Institute,
School of Computer Science,
Carnegie Mellon University, Pittsburgh, PA

**Comment by bhack** [comment]
Also in VOC 2007+2012 training experiment validation is missing so it is hard to control overfitting on training

**Comment by endernewton** [comment]
Does seem like incorporating an early stopping strategy will have to use a separate validation set, which we currently do not have.

**Comment by zacwellmer** [comment]
But if we do not adjust any of our hyper parameters or train using the test set shouldn't that be okay to use as a guide for early stopping?

**Comment by endernewton** [comment]
@zacwellmer Alrighty @bhack Ideally if we want to do testing, we should submit to the server. So strictly most AP numbers reported here and many papers are only on validation set. It is fine to use the validation set to guide early stopping, with the same reasoning that it is fine to use the validation set to tune options/hyper-parameters. However, ConvNets are data hungry, so I think (correct me if I am wrong) right now what people do for models that combine all voc 07/12 sets will first determine the hyper-parameters from a smaller set (e.g. using voc 07 test), and then fix those parameters and just train on all sets (including voc 07 test).

**Comment by bhack** [comment]
Yes meanly eary stopping is stop to learn (fine tune on train dataset) when you see that model generalization capability  is becoming worse. The problem with the test set is that you could test only the final model. This why submission to servers it is limited.. to avoid overfitting. At some point If you don't change model capacity and all the manual hyperparamters you can learn on test+validation but it is de facto another dataset so you really don't know precisely at what iteration stop.

**Comment by bhack** [comment]
Take a look also at [Early Stopping without a Validation Set](https://arxiv.org/abs/1703.09580)

**Comment by zacwellmer** [comment]
I'm assuming when you speak about only being able to test the final model you are referring to actually using this in a competition. Because the VOCdevkit does contain an image test set(haven't tried using this at all with COCO). If it's not possible to use the test set to check for early stopping, couldn't you just segment out a piece of the training set to use? Is there any reason this would be a bad idea aside from shrinking the size of your training set? Regardless, I will still look into early stopping without a validation set.

**Comment by zacwellmer** [comment]
I have implemented early stopping w/ a validation set for pascal_voc [here](https://github.com/zacwellmer/tf-faster-rcnn/tree/early_stop). However, we are reaching are only reaching ~.6 mAP after 70k iterations. The error is still decreasing so I'll continue the training process to see if we can get closer to the ~.7 hit in the paper. Any idea as to why we could be receiving such drastically different scores? Could it be because the graph is being reset after each early stop evaluation point?
